micro-rdk/src/common/webrtc/udp_mux.rs

```diff
@@ -0,0 +1,578 @@
+        if !state.interest {
+            state.interest = true;
+            Some(UdpMux {
+                muxer: self.clone(),
+                direction: STUN,
+                peer_addr: None,
+            })
+        } else {
+            None
+        }
+    }
+    pub(crate) fn get_dtls_mux(&self) -> Option<UdpMux> {
+        let state = &mut self.mux.lock().unwrap()[DTLS];
+        if !state.interest {
+            state.interest = true;
+            Some(UdpMux {
+                muxer: self.clone(),
+                direction: DTLS,
+                peer_addr: None,
+            })
+        } else {
+            None
+        }
+    }
+    async fn recv_from(&self, dir: usize, buf: &mut [u8]) -> Result<(usize, SocketAddr)> {
+        loop {
+            let r = match self.peek() {
+                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => (0, 0),
+                Ok(s) => s,
+                Err(e) => return Err(e),
+            };
+            if r.0 != 0 {
+                if dir == r.1 {
+                    let socket = self.socket.as_ref().get_ref();
+                    return socket.recv_from(buf);
+                }
+                if self.yield_or_discard(r.1, r.0)? {
+                    continue;
+                }
+            }
+            self.readable_udp_muxer(dir).await?;
+        }
+    }
+
+    fn register_waker(&self, waker: Waker, dir: usize) {
+        let mux = &mut self.mux.lock().unwrap()[dir];
+        if let Some(w) = mux.waker.take() {
+            if w.will_wake(&waker) {
+                mux.waker = Some(w);
+                return;
+            }
+            w.wake();
+        }
+        mux.waker = Some(waker);
+    }
+    fn deregister_waker(&self, dir: usize) {
+        let mux = &mut self.mux.lock().unwrap()[dir];
+        let _ = mux.waker.take();
+    }
+
+    fn yield_or_discard(&self, dir: usize, _len: u16) -> Result<bool> {
+        let socket = self.socket.as_ref().get_ref();
+        let mux = &mut self.mux.lock().unwrap()[dir];
+        if !mux.interest {
+            let mut buf = [0_u8; 1];
+            let _ = socket.recv_from(&mut buf)?;
+            return Ok(true);
+        }
+        if let Some(w) = mux.waker.take() {
+            w.wake();
+            return Ok(false);
+        }
+        Ok(false)
+    }
+
+    fn peek(&self) -> Result<(u16, usize)> {
+        let socket = self.socket.as_ref().get_ref();
+        let mut buf = [0_u8; 13];
+        let (len, _) = socket.peek_from(&mut buf)?;
+        if len != 13 {
+            let _ = socket.recv_from(&mut buf)?;
+            return Ok((0, 0));
+        }
+        Ok(self.read_header(buf))
+    }
+
+    fn read_header(&self, hdr: [u8; 13]) -> (u16, usize) {
+        let msg_type = hdr[0];
+        if msg_type < 2 {
+            // stun message
+            let len: u16 = u16::from_be_bytes(hdr[2..4].try_into().unwrap());
+            (len, STUN)
+        } else {
+            // assume DTLS record
```

Comment:
we currently only receive STUN or DTLS in the socket but potentially as we expand we may get other packets type

Reply:
ok, could we do a match statement with a case for `_ => unreachable!()` ? Feels less fragile

Reply:
To clarify further, assuming that any message type 2 or greater is DTLS automatically feels like a code smell to me because it relies on an external assumption. Wouldn't it be better to raise a specific error here if for some reason we receive a message through the UDP socket of an unexpected packet type?

Reply:
I see you point but it's actually ok. If the packet is not a proper DTLS record it will be rejected by the DTLS layer. 
We shouldn't report an error anyway and rather drop the packet silently. I don't want to write a state machine or a filter that supports identify correct STUN or DTLS packet it feels a bit cumbersome in this scenario